federatedscope/core/worker/server.py

```diff
@@ -178,9 +178,30 @@ def run(self):
             self.msg_handlers[msg.msg_type](msg)
 
         # Running: listen for message (updates from clients), aggregate and broadcast feedbacks (aggregated model parameters)
-        while self.state <= self.total_round_num:
-            msg = self.comm_manager.receive()
-            self.msg_handlers[msg.msg_type](msg)
+        min_received_num = self._cfg.asyn.min_received_num if hasattr(self._cfg, 'asyn') else self._cfg.federate.sample_client_num
+        with Timeout(self._cfg.asyn.timeout) as time_counter:
+            while self.state <= self.total_round_num:
+                try:
+                    msg = self.comm_manager.receive()
+                    move_on_flag = self.msg_handlers[msg.msg_type](msg)
+                    if move_on_flag:
+                        time_counter.reset()
+                except TimeoutError:
+                    print('Time out!')
+                    move_on_flag_eval = self.check_and_move_on(min_received_num=min_received_num, check_eval_result=True)
+                    move_on_flag = self.check_and_move_on(min_received_num=min_received_num)
+                    if not move_on_flag and not move_on_flag_eval:
+                        ## Time out, broadcast the model para and re-start the training round
+                        logging.info(
+                            '----------- Re-starting the training round (Round #{:d}) -------------'
+                                .format(self.state))
+                        # Clean the msg_buffer
+                        self.msg_buffer['train'][self.state].clear()
```

Comment:
Will the normal clients repeat local training and sending the same local parameters (since the global model and hyperparameters doesn't change) for the absence of some other clients?
Maybe we should set a maximum number of attempts, and report the failure of fl training after that.

Reply:
Here we restart the training round, including sample involved clients, local training, and then perform aggregation. So a client might repeat local training if it was sampled again.
We are going to implement a "cache" strategy to prevent redundant training, and it is a good idea to leave the failure exit here. Thanks!